python/src/iceberg/types.py

```diff
@@ -135,75 +179,122 @@ def doc(self) -> Optional[str]:
-            f"name={repr(self._name)}, field_type={repr(self._type)}, doc={repr(self._doc)})"
-        )
 
-    def __str__(self):
-        return (
-            f"{self._id}: {self._name}: {'optional' if self._is_optional else 'required'} {self._type}" ""
-            if self._doc is None
-            else f" ({self._doc})"
+class StructType(IcebergType):
+    """A struct type in Iceberg
+
+    Example:
+        >>> StructType(
+                NestedField(True, 1, "required_field", StringType()),
+                NestedField(False, 2, "optional_field", IntegerType())
         )
+    """
 
-    def __eq__(self, other):
-        if type(self) is type(other):
-            return (
-                self.is_optional == other.is_optional
-                and self.field_id == other.field_id
-                and self.name == other.name
-                and self.doc == other.doc
-                and self.type == other.type
-            )
-        return False
+    _instances: Dict[Tuple[NestedField, ...], "StructType"] = {}
 
+    def __new__(cls, *fields: NestedField):
+        cls._instances[fields] = cls._instances.get(fields) or object.__new__(cls)
+        return cls._instances[fields]
 
-class StructType(Type):
-    def __init__(self, fields: list):
+    def __init__(self, *fields: NestedField):
         super().__init__(
             f"struct<{', '.join(map(str, fields))}>",
-            f"StructType(fields={repr(fields)})",
+            f"StructType{repr(fields)}",
         )
         self._fields = fields
 
     @property
-    def fields(self) -> list:
+    def fields(self) -> Tuple[NestedField, ...]:
         return self._fields
 
-    def __eq__(self, other):
-        if type(self) is type(other):
-            return self.fields == other.fields
-        return False
 
+class ListType(IcebergType):
+    """A list type in Iceberg
+
+    Example:
+        >>> ListType(element_id=3, element_type=StringType(), element_is_optional=True)
+        ListType(element=NestedField(is_optional=True, field_id=3, name='element', field_type=StringType(), doc=None))
+    """
+
+    _instances: Dict[Tuple[bool, int, IcebergType], "ListType"] = {}
+
+    def __new__(
+        cls,
+        element_id: int,
+        element_type: IcebergType,
+        element_is_optional: bool,
+    ):
+        key = (element_is_optional, element_id, element_type)
+        cls._instances[key] = cls._instances.get(key) or object.__new__(cls)
+        return cls._instances[key]
 
-class ListType(Type):
-    def __init__(self, element: NestedField):
-        super().__init__(f"list<{element.type}>", f"ListType(element={repr(element)})")
-        self._element_field = element
+    def __init__(
+        self,
+        element_id: int,
+        element_type: IcebergType,
+        element_is_optional: bool,
+    ):
+        super().__init__(
+            f"list<{element_type}>",
+            f"ListType(element_is_optional={element_is_optional}, element_id={element_id}, "
```

Comment:
Minor: It would be nice if the argument order were consistent. This is slightly misleading because `element_is_optional` is last if you're passing by position.